src/plugins/content_management/server/core/msearch.ts

```diff
@@ -56,14 +60,31 @@ export class MSearchService {
     });
 
     const savedObjectsClient = await this.deps.getSavedObjectsClient();
+    const listingLimit = await this.deps.getConfig.listingLimit();
+    const defaultPerPage = await this.deps.getConfig.perPage();
+
+    const page = query.cursor ? Number(query.cursor) : 1;
+    const perPage = query.limit ? query.limit : defaultPerPage;
+
+    if (page * perPage >= listingLimit) {
+      throw new Error(
+        `Requested page ${page} with ${perPage} items per page exceeds the maximum allowed limit of ${listingLimit} items`
+      );
+    }
+
     const soResult = await savedObjectsClient.find({
       type: soSearchTypes,
+
       search: query.text,
       searchFields: [`title^3`, `description`, ...additionalSearchFields],
       defaultSearchOperator: 'AND',
-      // TODO: tags
-      // TODO: pagination
-      // TODO: sort
+
+      page,
+      perPage,
+
+      // tags
+      hasReference: query.tags?.included?.map((tagId) => ({ type: 'tag', id: tagId })),
```

Comment:
I know I hardcoded too the "tag" type but shouldn't we use a constant for it better?
Or maybe a `tagIdToSavedObjectReference` util?